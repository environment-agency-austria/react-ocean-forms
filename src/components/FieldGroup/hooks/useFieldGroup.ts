/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { useCallback, useMemo } from 'react';

import { useFormContext, useValidation, IValidationArgs, IBasicValidationState } from '../../../hooks';
import { useFullName, useFieldRegistration } from '../../../hooks/internal';
import { IFormContext, IFieldValues } from '../../FormContext';

import { getGroupValue } from './useFieldGroup.utils';
import { IUseFieldGroupArgs, IUseFieldGroupResult } from './useFieldGroup.types';

/**
 * @hidden
 */
export function useFieldGroup<TFieldValue = unknown>(props: IUseFieldGroupArgs<TFieldValue>): IUseFieldGroupResult {
  const formContext = useFormContext();

  const {
    name,
    label,
    defaultValues,
    values,
    disabled = formContext.disabled,
    plaintext = formContext.plaintext,
    asyncValidateOnChange = formContext.asyncValidateOnChange,
  } = props;

  const fullName = useFullName(name);
  const { validationState, validate, resetValidation, updateValidationState } = useValidation(props);

  /**
   * Triggers the validation of the group
   */
  const validateGroup = useCallback(
    (args?: Partial<IValidationArgs>): Promise<IBasicValidationState> => {
      const groupValue = getGroupValue<TFieldValue>(formContext, fullName);
      return validate(groupValue, args);
    },
    [formContext, fullName, validate],
  );

  // Register the group in the formContext, so the group
  // validation can be called on form submit.
  const registerFieldState = useMemo(() => ({
    label,
    isGroup: true,
    updateValidation: updateValidationState,
    validate: validateGroup,
    reset: resetValidation,
    getValue: () => ({}),
  }), [label, resetValidation, updateValidationState, validateGroup]);
  useFieldRegistration(
    fullName,
    registerFieldState,
  );

  /**
   * Listens to child field events, triggers validation if
   * needed and passes them to the higher context
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  const notifyFieldEvent = useCallback(
    (name: string, event: string, args?: unknown): void => {
      formContext.notifyFieldEvent(name, event, args);

      if (event !== 'change' && event !== 'blur') { return; }

      if (event === 'change') {
        const localName = name.substring(fullName.length + 1);

        const currentGroupValue = getGroupValue<TFieldValue>(formContext, fullName);
        const intermediateGroupValue = {
          ...(currentGroupValue === undefined ? { } : currentGroupValue),
          ...{
            // Override the value of the event sender, because
            // the Field didn't update its state yet, making the
            // Form.getValues() returning an old Field value.
            [localName]: args,
          },
        };

        void validate(
          intermediateGroupValue as TFieldValue,
          { checkAsync: asyncValidateOnChange },
        );
      } else if (!asyncValidateOnChange) {
        void validate(getGroupValue(formContext, fullName));
      }
    },
    [asyncValidateOnChange, formContext, fullName, validate],
  );

  const subContext: IFormContext = useMemo(() => ({
    ...formContext,
    fieldPrefix: fullName,
    notifyFieldEvent,
    disabled,
    plaintext,
    defaultValues: (defaultValues === undefined ? formContext.defaultValues : { ...formContext.defaultValues, ...{ [fullName]: defaultValues }}) as Partial<IFieldValues>,
    values: (values === undefined ? formContext.values : { ...formContext.values, ...{ [fullName]: values }}) as Partial<IFieldValues>,
  }), [defaultValues, disabled, formContext, fullName, notifyFieldEvent, plaintext, values]);

  const groupState = useMemo(
    () => ({
      fullName,
      isValidating: validationState.isValidating,
      isRequired: validationState.isRequired,
      valid: validationState.valid,
      error: validationState.error,
    }),
    [
      fullName,
      validationState,
    ],
  );

  return {
    groupFormContext: subContext,
    renderParams: groupState,
  };
}
