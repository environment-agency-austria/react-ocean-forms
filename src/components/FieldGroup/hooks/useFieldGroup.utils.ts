/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { IFormContext } from '../../FormContext';

/**
 * Helper function to get the correct value
 * of the group (including all values of the nested fields)
 * @param formContext Form context
 * @param fullName Full name of the field group
 * @hidden
 */
export function getGroupValue<TFieldValue extends {}>(formContext: IFormContext, fullName: string): TFieldValue | undefined {
  const formValues = formContext.getValues();

  const formValue = formValues[fullName];
  if (formValue === '' || formValue === undefined) {
    return undefined;
  }

  return formValue as TFieldValue;
}
