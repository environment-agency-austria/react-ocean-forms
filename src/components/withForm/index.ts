export * from './withForm';
export * from './withForm.types';
