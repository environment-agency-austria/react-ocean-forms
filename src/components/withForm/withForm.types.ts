/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module withForm
 */
import { IFormContext } from '../FormContext';

/**
 * Props for consumers of withForm props
 */
export interface IFormContextProps {
  /**
   * Form context of the parent form
   */
  context: IFormContext;
}
