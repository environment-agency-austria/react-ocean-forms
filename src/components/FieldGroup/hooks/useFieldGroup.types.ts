import { IUseValidationArgs, IValidationState } from '../../../hooks';
import { IFormContext } from '../../FormContext';

/**
 * Props for the field group component
 */
export interface IUseFieldGroupArgs<TFieldValue = unknown> extends IUseValidationArgs<TFieldValue> {
  /**
   * Message id of the label that will be displayed along the input.
   * If you don't want to use any i18n features you can pass a raw message instead.
   */
  label: string;
  /**
   * If set to true the form will trigger asynchronous validation on Fields whenever
   * they change (e.g. on key press). Default behaviour is that the fields will only
   * async validate when they loose focus.
   * @default Form.asyncValidateOnChange
   */
  asyncValidateOnChange?: boolean;
  /**
   * Overwrites the Form default values for the child fields of this field group.
   * Those values will be put into the according fields when the form initializes.
   */
  defaultValues?: TFieldValue;
  /**
   * Overwrites the Form values for the child fields of this field group. Changing
   * this property will update all Field values, overwriting their default values
   * but also any value the user put in.
   */
  values?: TFieldValue;
  /**
   * Overwrites the disabled state for this field group.
   * @default Form.disabled
   */
  disabled?: boolean;
  /**
   * Overwrites the plaintext state for this field group.
   * @default Form.plaintext
   */
  plaintext?: boolean;
}

/**
 * Meta information about the group
 * for the render prop
 */
export interface IFieldGroupRenderParams extends IValidationState {
  /**
   * Full name of the group
   */
  fullName: string;
}

/**
 * Result of the useFieldGroup hook
 */
export interface IUseFieldGroupResult {
  /**
   * Overriden form context for the group
   */
  groupFormContext: IFormContext;
  /**
   * Params with meta information to be passed
   * to the render method of FieldGroup
   */
  renderParams: IFieldGroupRenderParams;
}
