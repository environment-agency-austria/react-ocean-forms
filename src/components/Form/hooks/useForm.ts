/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module Form
 */
import { useMemo, useState, useCallback } from 'react';

import {
  stringFormatter as defaultStringFormatter,
  noopFunction,
  parseValidationError,
  getDeepValue,
} from '../../../utils';
import { IBasicValidationState } from '../../../hooks';
import { IFormContext, IFieldValues } from '../../FormContext';
import { IFormProps } from '../Form.types';
import { useFieldEvents, useFieldStates, useIsUnmounted } from '../../../hooks/internal';

/**
 * @hidden
 */
export function useForm<TFieldValues extends Record<string, unknown> = IFieldValues>(
  props: IFormProps<TFieldValues>
): IFormContext<TFieldValues> {
  const [busyState, setBusyState] = useState(false);

  const {
    defaultValues = {},
    values,
    asyncValidationWait = 400,
    asyncValidateOnChange = false,
    formatString = defaultStringFormatter,
    disabled = false,
    plaintext = false,
    busy = busyState,
    resetOnSubmit,
    onReset = noopFunction,
    onSubmit = noopFunction,
    onValidate = noopFunction,
  } = props;

  const { registerListener, unregisterListener, notifyListeners } = useFieldEvents();
  const { getFieldState, registerField, unregisterField, forEachFieldState } = useFieldStates();
  const isUnmounted = useIsUnmounted();

  /**
   * Generates and returns an object that contains
   * all values from all the fields.
   * @returns Current values in form of { name: value, name2: value2, ... }
   */
  const getValues = useCallback((): TFieldValues => {
    const values = {} as TFieldValues;

    forEachFieldState((state, name) => {
      if (state.isGroup === true) {
        return;
      }

      const nameParts = name.split('.');
      let valueRef: Record<string, unknown> = values;

      nameParts.forEach((key, index) => {
        if (nameParts.length === 1 || index === nameParts.length - 1) {
          valueRef[key] = state.getValue();
        } else {
          if (valueRef[key] === undefined) {
            valueRef[key] = {};
          }
          valueRef = valueRef[key] as TFieldValues;
        }
      });
    });

    return values;
  }, [forEachFieldState]);

  /**
   * Sets the state of all fields back
   * to the default state.
   */
  const reset = useCallback((): void => {
    forEachFieldState(state => {
      state.reset();
    });

    onReset();
  }, [forEachFieldState, onReset]);

  /**
   * Triggers the form wide validation callback if given
   * @returns Validation state of the form
   */
  const triggerFormValidation = useCallback((): boolean => {
    const values = getValues();
    const result = onValidate(values);

    // If the callback returned null then the form is valid
    if (result === null || result === undefined) {
      return true;
    }

    // Otherwise parse the result object and update the
    // field states.
    let allFieldsValid = true;
    forEachFieldState((state, name) => {
      const fieldError = parseValidationError(getDeepValue(name, result));
      const isValid = fieldError === null || typeof fieldError !== 'object';

      if (isValid) {
        return;
      }

      state.updateValidation({
        valid: false,
        error: fieldError,
      });
      allFieldsValid = false;
    });

    return allFieldsValid;
  }, [forEachFieldState, getValues, onValidate]);

  /**
   * Submits the form - triggers any validations if needed
   * and raises the onFormSubmit prop callback if the
   * form is currently valid.
   * @param submitArgs Arguments that will be passed
   * to the onSubmit callback
   */
  const submit = useCallback(
    async (submitArgs?: unknown): Promise<void> => {
      setBusyState(true);

      // Iterate through all fields and validate them
      // if needed.
      const validations: Promise<IBasicValidationState>[] = [];
      forEachFieldState(field => {
        validations.push(
          field.validate({
            checkAsync: true,
            immediateAsync: true,
          })
        );
      });
      const validationStates = await Promise.all(validations);

      const notifyInvalid = (): void => {
        notifyListeners('_form', 'submit-invalid');
        setBusyState(false);
      };

      // Check if all fields are valid
      const allValid = validationStates.every(state => state.valid === true);
      if (allValid === false) {
        notifyInvalid();
        return;
      }

      // Call the form wide validation
      const formValid = triggerFormValidation();
      if (formValid === false) {
        notifyInvalid();
        return;
      }

      // Await the result from callOnSubmit
      const callOnSubmitResult = onSubmit(getValues(), submitArgs);

      // Make sure the state is cleaned up before
      const cleanup = (resetForm: boolean | undefined): void => {
        if (isUnmounted.current) return;

        setBusyState(false);
        if (resetForm) {
          reset();
        }
      };
      if (callOnSubmitResult instanceof Promise) {
        void callOnSubmitResult.then(() => {
          cleanup(resetOnSubmit);
        });
      } else {
        cleanup(resetOnSubmit);
      }
    },
    [
      forEachFieldState,
      getValues,
      isUnmounted,
      notifyListeners,
      onSubmit,
      reset,
      resetOnSubmit,
      triggerFormValidation,
    ]
  );

  const formContext: IFormContext<TFieldValues> = useMemo(
    () => ({
      fieldPrefix: null,
      defaultValues,
      values,
      asyncValidationWait,
      stringFormatter: formatString,
      disabled,
      plaintext,
      busy,
      asyncValidateOnChange,

      registerField,
      unregisterField,

      notifyFieldEvent: notifyListeners,
      registerListener,
      unregisterListener,

      getFieldState: getFieldState,
      getValues: getValues,

      reset,
      submit,
    }),
    [
      asyncValidateOnChange,
      asyncValidationWait,
      busy,
      defaultValues,
      disabled,
      formatString,
      getFieldState,
      getValues,
      notifyListeners,
      plaintext,
      registerField,
      registerListener,
      reset,
      submit,
      unregisterField,
      unregisterListener,
      values,
    ]
  );

  return formContext;
}
