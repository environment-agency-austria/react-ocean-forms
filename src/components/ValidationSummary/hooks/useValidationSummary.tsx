/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module ValidationSummary
 */
import React, { useRef, useCallback, useState } from 'react';

import { useFormContext, IValidationState, useFormEventListener } from '../../../hooks';
import { FieldError } from '../../FieldError';

interface IValidationEventArgs extends IValidationState {
  label: string;
}

/**
 * @hidden
 */
export interface IInvalidField {
  id: string;
  name: string;
  error: React.ReactNode;
  linkCallback(event: React.MouseEvent): void;
}

/**
 * @hidden
 */
export interface IUseValidationSummaryResult {
  headerRef: React.MutableRefObject<HTMLHeadingElement | null>;
  errorList: [string, IInvalidField][];
}

/**
 * Custom hook for the validation summary logic
 * @param id Id of the validation summary
 * @param disableFocusOnSubmit If set to true the validation summary will stop automatically scrolling to itself when the user clicks on a submit button and the form is invalid.
 * @hidden
 */
export function useValidationSummary(id: string, disableFocusOnSubmit: boolean): IUseValidationSummaryResult {
  const headerRef = useRef<HTMLHeadingElement | null>(null);
  const [errorList, setErrorList] = useState<[string, IInvalidField][]>([]);
  const { stringFormatter } = useFormContext();

  const handleEvent = useCallback((name: string, event: string, args?: unknown): void => {
    if (event === 'validation') {
      const validationEventArgs = args as IValidationEventArgs;

      // If the field has been reported as valid delete it from our error map
      if (validationEventArgs.valid || validationEventArgs.error === null) {
        setErrorList((oldList) => oldList.filter(([oldName]) => name !== oldName));
        return;
      }

      // Create the individual error renderers
      const renderedErrors = (
        <FieldError
          id={name}
          invalid={!validationEventArgs.valid}
          error={validationEventArgs.error}
        />
      );

      const linkCallback = (event: React.MouseEvent): void => {
        event.preventDefault();
        const input = document.getElementById(name);
        if (input !== null) {
          input.focus();
        }
      };

      // Create the meta information for the error map
      setErrorList((oldList) => {
        return [
          ...oldList.filter(([oldName]) => name !== oldName),
          [
            name,
            {
              id: name,
              name: stringFormatter(validationEventArgs.label),
              error: renderedErrors,
              linkCallback,
            },
          ]
        ];
      });
    } else if (event === 'submit-invalid' && headerRef.current !== null && !disableFocusOnSubmit) {
      headerRef.current.scrollIntoView({ block: 'start', behavior: 'smooth' });
    }
  }, [disableFocusOnSubmit, stringFormatter]);

  useFormEventListener(id, handleEvent);

  return { headerRef, errorList };
}
