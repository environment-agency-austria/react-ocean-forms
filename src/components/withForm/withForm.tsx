/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { useFormContext } from '../../hooks';
import { IFormContextProps } from './withForm.types';

/**
 * Internal type for form component props. Deals with the
 * defaultProp logic and removes the form context props.
 */
type FormComponentProps<TComp> = Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IFormContextProps>;

/**
 * High order component for consuming the form context
 * @deprecated Deprecated in favour of `useFormContext` hook
 */
export const withForm = <TComp extends React.ComponentType<TProps>, TProps extends IFormContextProps = PropsOf<TComp>>(Component: TComp):
React.ComponentType<FormComponentProps<TComp>> => {
  /**
   * We need to cast the component back to a
   * React.ComponentType in order to use it in
   * the FormContext.Consumer. Otherwise
   * typescript will cry about the component
   * not having a constructor / call signatures.
   */
  const CastedComponent = Component as React.ComponentType<TProps>;

  /**
   * Component that injects the form context prop
   * to the wrapped component
   */
  const FormComponent: React.SFC<FormComponentProps<TComp>> = (props: FormComponentProps<TComp>): JSX.Element => {
    const context = useFormContext();

    // @ts-ignore
    return <CastedComponent {...props} context={context} />;
  };
  FormComponent.displayName = `FormComponent(${getDisplayName(Component)})`;

  return FormComponent;
};
