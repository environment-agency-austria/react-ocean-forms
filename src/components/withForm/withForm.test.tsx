import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { createMockFormContext } from '../../test-utils/enzymeFormContext';

import { useFormContext } from '../../hooks';
import { withForm } from './withForm';
import { IFormContextProps } from './withForm.types';

jest.mock('../../hooks');

describe('withForm', () => {
  const formContext = createMockFormContext();
  (useFormContext as jest.Mock).mockReturnValue(formContext);

  const TestComponent = (): JSX.Element => <div id="test-component" />;
  const WrappedComponent = withForm(TestComponent);

  const setup = (props?: Partial<IFormContextProps>): ShallowWrapper =>
    shallow(<WrappedComponent {...props} />);
  const wrapper = setup();

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('FormContext.Provider', () => {
    const childWrapper = wrapper.dive();

    it('should render the children without error', () => {
      expect(childWrapper).toMatchSnapshot();
    });

    it('should have the formContext supplied as a prop', () => {
      expect(wrapper.prop('context')).toBe(formContext);
    });
  });
});
