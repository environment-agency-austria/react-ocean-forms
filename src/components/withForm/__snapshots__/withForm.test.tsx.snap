// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withForm FormContext.Provider should render the children without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Consumer>
    [Function]
  </Consumer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "context": Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction],
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "context": Object {
          "asyncValidateOnChange": false,
          "asyncValidationWait": 400,
          "busy": false,
          "defaultValues": Object {},
          "disabled": false,
          "fieldPrefix": null,
          "getFieldState": [MockFunction],
          "getValues": [MockFunction],
          "notifyFieldEvent": [MockFunction],
          "plaintext": false,
          "registerField": [MockFunction],
          "registerListener": [MockFunction],
          "stringFormatter": [MockFunction],
          "submit": [MockFunction],
          "unregisterField": [MockFunction],
          "unregisterListener": [MockFunction],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withForm should render without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormComponent(TestComponent) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
